### PR TITLE
Update lightbox-photoswipe.php

### PR DESCRIPTION
**Pull request by B-e-n-G**
Added a new option `use_cache` in order to use the WordPress caching mechanisms [wp_cache_add](https://developer.wordpress.org/reference/functions/wp_cache_add/) and [wp_cache_get](https://developer.wordpress.org/reference/functions/wp_cache_get/) instead of the database table `wp_lightbox_photoswipe_img`.

This is working well with varios caching plugins like [Redis Object Cache](https://wordpress.org/plugins/redis-cache/).

The description of the option in the backend may need to be improved:
> Use wp_cache_add/get instead database

...and there aren't any translations yet.

### CODE REVIEW

**Comment by arnowelzel** [comment]
Interesting idea, looks good. The only thing I am missing yet is to remove the database tables if the cache is used:

When the cache is active, make sure the existing tables are removed similar to https://github.com/arnowelzel/lightbox-photoswipe/blob/98b53bd9fc46a8ca0c8ae65af4ce47f5df660481/uninstall.php#L9

If the cache works well it might also be worth it to omit the database use completely.  However it should be possible to clean up the cache manually as well - at least when removing the plugin (see https://github.com/arnowelzel/lightbox-photoswipe/blob/98b53bd9fc46a8ca0c8ae65af4ce47f5df660481/uninstall.php#L23). Without the cache one can just remove and reinstall the plugin to make sure all cached data is gone. Does this also happen when the plugin get's removed after it added some cache entries?

File: lightbox-photoswipe.php [hunk 1]
```diff
@@ -532,65 +535,98 @@ function outputCallbackProperties($matches)
...
-                    if (false !== $exif) {
-                        $exifCamera = $this->getExifCamera($exif);
-                        $exifFocal = $this->exifGetFocalLength($exif);
-                        $exifFstop = $this->exifGetFstop($exif);
-                        $exifShutter = $this->exifGetShutter($exif);
-                        $exifIso = $this->exifGetIso($exif);
-                        $exifDateTime = $this->exifGetDateTime($exif);
+            if ( $this->use_cache ) {
+                $cache_key = "img:$file";
+
```
Comment (arnowelzel) [inline lightbox-photoswipe.php:540]: The file alone is not sufficient to determine the cache key. There is a reason why I used the filename and date to build the key - when a file gets updated, the key must change as well, even if the name stays unchanged.

Also "img:" is not ideal. Better use "lbwps:" similar to all other places where identifiers are used in "public" places (frontend script, options passed to the frontend etc.).

So keep the code to determine the key above this code block and then use the key based on file name and modification date:

`
$imgdate = @filemtime($file);
$imgdate = 0;
$imgkey = md5($file) . '-'. $imgdate;

if ( $this->use_cache ) {
    $cache_key = "lbwps:$imgkey";
`

File: lightbox-photoswipe.php [hunk 2]
```diff
@@ -532,65 +535,98 @@ function outputCallbackProperties($matches)
...
-                        $exifDateTime
-                    );
-                    $wpdb->query($sql);
+                extract( $imgDetails );
+            } else {
+                $imgdate = @filemtime($file);
+                if (false == $imgdate) {
+                    $imgdate = 0;
+                }
+                $imgkey = md5($file) . '-'. $imgdate;
```
Comment (arnowelzel) [inline lightbox-photoswipe.php:574]: See above - this code block should be used for cases with/without cache.

Comment (arnowelzel) [inline lightbox-photoswipe.php:574]: See above - this code block should be used for cases with/without cache.

File: lightbox-photoswipe.php (hunk 1 above)
Comment (arnowelzel) [inline lightbox-photoswipe.php:540]: The file alone is not sufficient to determine the cache key. There is a reason why I used the filename and date to build the key - when a file gets updated, the key must change as well, even if the name stays unchanged.

Also "img:" is not ideal. Better use "lbwps:" similar to all other places where identifiers are used in "public" places (frontend script, options passed to the frontend etc.).

So keep the code to determine the key above this code block and then use the key based on file name and modification date:

```
$imgdate = @filemtime($file);
$imgdate = 0;
$imgkey = md5($file) . '-'. $imgdate;

if ( $this->use_cache ) {
    $cache_key = "lbwps:$imgkey";
```

File: lightbox-photoswipe.php [hunk 3]
```diff
@@ -535,8 +537,14 @@ function outputCallbackProperties($matches)
...
 
+            $imgMtime = @filemtime( $file );
+            if ( false === $imgMtime ) {
+                $imgMtime = 0;
+            }
+            $imgkey = hash( 'md5', $file . $imgMtime );
+
             if ( $this->use_cache ) {
-                $cache_key = "img:$file";
+                $cache_key = "img:$imgkey";
```
Comment (arnowelzel) [inline lightbox-photoswipe.php:547]: I'd prefer `"lbwps:$imgkey"` since `img:` is quite generic and may be used by other components as well.

**Comment by arnowelzel** [comment]
Thank you for the quick fixes! The only little thing left is the cache key - see my comment. Otherwise it looks good.

**Comment by B-e-n-G** [comment]
> 
> 
> Interesting idea, looks good. The only thing I am missing yet is to remove the database tables if the cache is used:
> 
> When the cache is active, make sure the existing tables are removed similar to
> 
> https://github.com/arnowelzel/lightbox-photoswipe/blob/98b53bd9fc46a8ca0c8ae65af4ce47f5df660481/uninstall.php#L9
> 
> If the cache works well it might also be worth it to omit the database use completely. However it should be possible to clean up the cache manually as well - at least when removing the plugin (see
> 
> https://github.com/arnowelzel/lightbox-photoswipe/blob/98b53bd9fc46a8ca0c8ae65af4ce47f5df660481/uninstall.php#L23
> ). Without the cache one can just remove and reinstall the plugin to make sure all cached data is gone. Does this also happen when the plugin get's removed after it added some cache entries?

Good point... I have added an action `update_option_lightbox_photoswipe_use_cache` which is calling the new method `update_option_use_cache`. This method clears the table and the cron job in case of selected option "Use wp_cache_add..." and creates table and cron in case of this option isn't selected.

Clearing the redis cache in case of uninstall this plugin isn't necessary, because [wp_cache_add](https://developer.wordpress.org/reference/functions/wp_cache_add/) becomes `CACHE_EXPIRE_IMG_DETAILS` as third parameter. The redis will clear all entries after this time automatically

File: lightbox-photoswipe.php (hunk 1 above)
Comment (B-e-n-G) [reply lightbox-photoswipe.php]: Hmm ok, but I have used "lbwps" as third parameter of [wp_cache_add](https://developer.wordpress.org/reference/functions/wp_cache_add/). The more professionals cache plugins (like [Redis Object Cache](https://wordpress.org/plugins/redis-cache/)) are supporting cache groups.

That means technically the redis key will be:
`<wp_prefix>:<group>:<mykey>`

...for example:
`wp_:lbwps:img:441d5118199e24eeff8a1c3602bc7cbd`

Your suggestion `$cache_key = "lbwps:$imgkey";` will result in a redis key like:
`wp_:lbwps:lbwps:441d5118199e24eeff8a1c3602bc7cbd`

Comment (arnowelzel) [reply lightbox-photoswipe.php]: Ok, got it. I was not aware about the cache group parameter. Thanks for pointing this out!

File: lightbox-photoswipe.php (hunk 3 above)
Comment (arnowelzel) [reply lightbox-photoswipe.php:547]: I was not aware of the cache group parameter for `wp_cache_add()`, so no need to change anything.

**Pull request merged**